src/automatic_dantzig_wolfe.jl

```diff
@@ -0,0 +1,380 @@
+
+# Available scores that can be used in an automatic Dantzig-Wolfe decomposition
+# inaktive means that automatic Dantzig-Wolfe is not used
+@enum(AutoDwStrategy, inaktive, white_score, block_border_score, relative_border_area_score)
+
+# Decomposes the given JuMP Model automatically
+function automatic_dw_decomposition!(model::JuMP.Model)
+    model.ext[:decomposition_structure] = BlockDecomposition.get_best_block_structure(model)
+    decomposition_axis = BlockDecomposition.Axis(
+        1:length(model.ext[:decomposition_structure].blocks)
+    )
+    decomposition = BlockDecomposition.decompose_leaf(
+        model,
+        BlockDecomposition.DantzigWolfe,
+        decomposition_axis
+    )
+    BlockDecomposition.register_subproblems!(
+        decomposition,
+        decomposition_axis,
+        BlockDecomposition.DwPricingSp,
+        BlockDecomposition.DantzigWolfe
+    )
+    return
+end
+
+# Finds the best decomposition structure to be used by the solver
+function get_best_block_structure(model::JuMP.Model)
+    @assert(model.ext[:automatic_dantzig_wolfe] != inaktive)
+    block_structures = get_all_block_structures(model)
+    score_type = model.ext[:automatic_dantzig_wolfe]
+    if score_type == white_score
+        white_scores =  BlockDecomposition.white_scores(block_structures)
+        result = block_structures[argmax(white_scores)]
+    elseif score_type == block_border_score
+        block_border_scores = BlockDecomposition.block_border_scores(block_structures)
+        result = block_structures[argmax(block_border_scores)]
+    elseif score_type == relative_border_area_score
+        relative_border_area_scores = BlockDecomposition.relative_border_area_scores(block_structures)
+        result = block_structures[argmin(relative_border_area_scores)]
+    end
+    return result
+end
+
+# Returns all possible block structures of the given model
+function get_all_block_structures(model::JuMP.Model)
+    model_description = get_model_description(model)
+    block_structures = Array{BlockStructure,1}()
+    axesSets = collect(powerset(collect(model_description.axes)))
+    for axes in axesSets
+        block_structure0, block_structure1 = get_block_structure(axes, model_description, model)
+        # Only add structures that were not found already and where the set of master
+        # constraints is not empty
+        if !structure_exists(block_structures, block_structure0) &&
+                !isempty(block_structure0.master_constraints)
+            push!(block_structures, block_structure0)
+        end
+        if !structure_exists(block_structures, block_structure1) &&
+                !isempty(block_structure1.master_constraints)
+            push!(block_structures, block_structure1)
+        end
+    end
+    return block_structures
+end
+
+# Contains all the information about the constraints and variables in a model we might need
+# to compute different block structures
+mutable struct ModelDescription
```

Comment:
Are you sure it's mutable ?

Reply:
I changed to code so that is does not need to be mutable anymore. This makes more sense because the model description does not need to be modified once it is created. Thanks for the remark. I think the code is now better to read.